### ISSUE TITLE
RadosGW not ready due to high memory usage

### ISSUE THREAD

**Issue by TimVerboisCgk** [rook/rook#14881]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:** All RadosGW pods go in "not ready" state, since the readinessprobe times out:

```
rook-ceph-rgw-applicationpool-a-8fbb9f7c5-5h6jq           1/2     Running     0          24d
rook-ceph-rgw-applicationpool-a-8fbb9f7c5-dbwtd           1/2     Running     0          24d
rook-ceph-rgw-applicationpool-a-8fbb9f7c5-fdgzw           1/2     Running     0          24d
rook-ceph-rgw-applicationpool-a-8fbb9f7c5-ldhlm           1/2     Running     0          24d
rook-ceph-rgw-applicationpool-a-8fbb9f7c5-nvkqp           1/2     Running     0          24d
rook-ceph-rgw-applicationpool-a-8fbb9f7c5-q82gp           1/2     Running     0          24d
```
You see here that 1 out of 2 containers (istio) is not ready:

```
Containers:
  rgw:
    Container ID:  docker://ad8a949d073f32142310b461563301d6f1e7e061db8a1da978b0e935ea7d433d
    Image:         quay.io/ceph/ceph:v18.2.2
    Image ID:      docker-pullable://quay.io/ceph/ceph@sha256:f8d467dcf49d13b8ea42229d89be642581110175d8ce36e216aefc9b32b0854d
    Port:          <none>
    Host Port:     <none>
    Command:
      radosgw
    Args:
      --fsid=c2d6ff39-1dcb-4184-aac4-7b3fcef65a01
      --keyring=/etc/ceph/keyring-store/keyring
      --default-log-to-stderr=true
      --default-err-to-stderr=true
      --default-mon-cluster-log-to-stderr=true
      --default-log-stderr-prefix=debug
      --default-log-to-file=false
      --default-mon-cluster-log-to-file=false
      --mon-host=$(ROOK_CEPH_MON_HOST)
      --mon-initial-members=$(ROOK_CEPH_MON_INITIAL_MEMBERS)
      --id=rgw.applicationpool.a
      --setuser=ceph
      --setgroup=ceph
      --foreground
      --rgw-frontends=beast port=8080
      --host=$(POD_NAME)
      --rgw-mime-types-file=/etc/ceph/rgw/mime.types
      --rgw-realm=applicationpool-prd
      --rgw-zonegroup=applicationpool-prd
      --rgw-zone=applicationpool-prd-master
    State:          Running
      Started:      Fri, 27 Sep 2024 11:23:45 +0200
    Ready:          False
    Restart Count:  0
    Requests:
      memory:   1Gi
    Readiness:  exec [bash -c #!/usr/bin/env bash
```

So, this container (the rgw container) is marked as "not Ready"

**Expected behavior:** When memory limit is reached, or an oomkill should be done, or the pod process (radosgw) should crash, triggering a restart.

**How to reproduce it (minimal and precise):** We had the problem several times, we cleaned up a bucked with lots of parallel sessions.  When you start this, memory in the pod will go up fast:

<img width="1325" alt="image" src="https://github.com/user-attachments/assets/44f69cc2-1b17-4795-9bcb-d0c7a3d5806e">

At the highest point (flat line), the radosgw all went into "not-ready".  This is due to the fact that the readiness probe failed:

```
Warning  Unhealthy  4m36s (x308 over 49m)  kubelet  Readiness probe failed: command "bash -c #!/usr/bin/env 
<truncated>
...
timed out
```

to summarize: the readiness probe timed out.  Probably because it could not get any memory anymore.  This behavior looks like a Linux VM that runs out of memory, at some point it cannot even create 1 process anymore and it is completely stuck.  Only a reboot can help.

Running a `kubectl rollout restart` fixed the problem, but like I said, it happened several times before.

**File(s) to submit**:

* Cluster CR (custom resource): CephObjectStore

So, this is the CephObjectStore:

```yaml
apiVersion: ceph.rook.io/v1
kind: CephObjectStore
metadata:
  finalizers:
  - cephobjectstore.ceph.rook.io
  generation: 7
  labels:
    app.kubernetes.io/instance: clsp001-ceph
  name: applicationpool
  namespace: rook-ceph
spec:
  dataPool:
    erasureCoded:
      codingChunks: 1
      dataChunks: 2
    failureDomain: host
    mirroring: {}
    quotas: {}
    replicated:
      size: 0
    statusCheck:
      mirror: {}
  gateway:
    instances: 6
    placement:
      nodeAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
          nodeSelectorTerms:
          - matchExpressions:
            - key: storage-node
              operator: In
              values:
              - "true"
      tolerations:
      - key: storage-node
        operator: Exists
    port: 80
    resources:
      requests:
        memory: 1024Mi
    sslCertificateRef: null
  healthCheck: {}
  metadataPool:
    erasureCoded:
      codingChunks: 0
      dataChunks: 0
    failureDomain: host
    mirroring: {}
    quotas: {}
    replicated:
      size: 3
    statusCheck:
      mirror: {}
  preservePoolsOnDelete: true
  zone:
    name: applicationpool-prd-master
```

We are using istio.

**Logs to submit**:
The pods itself did not log anything unusual:

```
debug 2024-10-21T13:31:35.583+0000 7f9fd88ce700  1 ====== starting new request req=0x7f9f6b42d730 =====
debug 2024-10-21T13:31:35.583+0000 7fa07a211700  1 ====== req done req=0x7f9f6b42d730 op status=0 http_status=200 latency=0.000000000s ======
debug 2024-10-21T13:31:35.583+0000 7fa07a211700  1 beast: 0x7f9f6b42d730: 10.244.148.35 - obc-cin-bed-pil-mgmt-tntv2-s3 [21/Oct/2024:13:31:35.583 +0000] "POST /cin-bed-pil-mgmt.tntv2-418a1c4f-4258-4400-90ff-d7cfada7029c?delete HTTP/1.1" 200 137 - "aws-cli/2.18.10 md/awscrt#0.22.0 ua/2.0 os/linux#5.15.0-118-generic md/arch#x86_64 lang/python#3.12.6 md/pyimpl#CPython cfg/retry-mode#standard md/installer#docker md/distrib#amzn.2 md/prompt#off md/command#s3api.delete-objects" - latency=0.000000000s
debug 2024-10-21T13:31:35.735+0000 7f9fd88ce700  1 ====== starting new request req=0x7f9f69afb730 =====
debug 2024-10-21T13:31:35.755+0000 7f9fd88ce700  1 ====== starting new request req=0x7f9f6cada730 =====
debug 2024-10-21T13:31:35.767+0000 7f9fd88ce700  1 ====== req done req=0x7f9f6cada730 op status=0 http_status=200 latency=0.012000085s ======
debug 2024-10-21T13:31:35.767+0000 7f9fd88ce700  1 beast: 0x7f9f6cada730: 10.244.148.61 - obc-cin-bed-pil-mgmt-tntv2-s3 [21/Oct/2024:13:31:35.755 +0000] "GET /cin-bed-pil-mgmt.tntv2-418a1c4f-4258-4400-90ff-d7cfada7029c?list-type=2&max-keys=15&prefix=cin-bed-pil-icn_tntv2-spark%2F&continuation-token=cin-bed-pil-icn_tntv2-spark%2Fspark-upload-f178161a-248f-4ac5-aaa0-43b2a8d13d98%2Fpool-strategy.xml&encoding-type=url HTTP/1.1" 200 5492 - "aws-cli/2.18.10 md/awscrt#0.22.0 ua/2.0 os/linux#5.15.0-118-generic md/arch#x86_64 lang/python#3.12.6 md/pyimpl#CPython cfg/retry-mode#standard md/installer#docker md/distrib#amzn.2 md/prompt#off md/command#s3api.list-objects-v2" - latency=0.012000085s
debug 2024-10-21T13:31:36.227+0000 7f9fd88ce700  1 ====== starting new request req=0x7f9f6aea2730 =====
debug 2024-10-21T13:33:01.371+0000 7fa0886d0700  0 rgw bucket sync thread: check_bucket_shards: resharding needed: stats.num_objects=1353626 shard max_objects=1100000
debug 2024-10-21T13:33:01.371+0000 7fa0886d0700  1 rgw bucket sync thread: RGWRados::check_bucket_shards bucket cin-bed-prd-mgmt.tntv2-931d05c3-fe82-4270-a320-2085c04a347a needs resharding; current num shards 11; new num shards 109 (suggested 108)
```

This was in the `kubeport describe` of the gateway pods as an event.  It also turns up on the nodes messages file:

```
Warning  Unhealthy  4m36s (x308 over 49m)  kubelet  Readiness probe failed: command "bash -c #!/usr/bin/env bash\n\nPROBE_TYPE=\"readiness\"\nPROBE_PORT=\"8080\"\nPROBE_PROTOCOL=\"HTTP\"\n\n# standard bash codes start at 126 and progress upward. pick error codes from 125 downward for\n# script as to allow curl to output new error codes and still return a distinctive number.\nUSAGE_ERR_CODE=125\nPROBE_ERR_CODE=124\n# curl error codes: 1-123\n\nSTARTUP_TYPE='startup'\nREADINESS_TYPE='readiness'\n\nRGW_URL=\"$PROBE_PROTOCOL://0.0.0.0:$PROBE_PORT\"\n\nfunction check() {\n  local URL=\"$1\"\n  # --insecure - don't validate ssl if using secure port only\n  # --silent - don't output progress info\n  # --output /dev/stderr - output HTML header to stdout (good for debugging)\n  # --write-out '%{response_code}' - print the HTTP response code to stdout\n  curl --insecure --silent --output /dev/stderr --write-out '%{response_code}' \"$URL\"\n}\n\nhttp_response=\"$(check \"$RGW_URL\")\"\nretcode=$?\n\nif [[ $retcode -ne 0 ]]; then\n  # if this is the startup probe, always returning failure. if startup probe passes, all subsequent\n  # probes can rely on the assumption that the health check was once succeeding without errors.\n  # if this is the readiness probe, we know that curl was previously working correctly in the\n  # startup probe, so curl error most likely means some new error with the RGW.\n  echo \"RGW health check failed with error code: $retcode. the RGW likely cannot be reached by clients\" >/dev/stderr\n  exit $retcode\nfi\n\nRGW_RATE_LIMITING_RESPONSE=503\nRGW_MISCONFIGURATION_RESPONSE=500\n\nif [[ $http_response -ge 200 ]] && [[ $http_response -lt 400 ]]; then\n  # 200-399 are successful responses. same behavior as Kubernetes' HTTP probe\n  exit 0\n\nelif [[ $http_response -eq $RGW_RATE_LIMITING_RESPONSE ]]; then\n  # S3's '503: slow down' code is not an error but an indication that RGW is throttling client\n  # traffic. failing the readiness check here would only cause an increase in client connections on\n  # other RGWs and likely cause those to fail also in a cascade. i.e., a special healthy response.\n  echo \"INFO: RGW is rate limiting\" 2>/dev/stderr\n  exit 0\n\nelif [[ $http_response -eq $RGW_MISCONFIGURATION_RESPONSE ]]; then\n  # can't specifically determine if the RGW is running or not. most likely a misconfiguration.\n  case \"$PROBE_TYPE\" in\n  \"$STARTUP_TYPE\")\n    # fail until we can accurately get a valid healthy response when runtime starts.\n    echo 'FAIL: HTTP code 500 suggests an RGW misconfiguration.' >/dev/stderr\n    exit $PROBE_ERR_CODE\n    ;;\n  \"$READINESS_TYPE\")\n    # config likely modified at runtime which could result in all RGWs failing this check.\n    # occasional client failures are still better than total failure, so ignore this\n    echo 'WARN: HTTP code 500 suggests an RGW misconfiguration' >/dev/stderr\n    exit 0\n    ;;\n  *)\n    # prior arg validation means this path should never be activated, but keep to be safe\n    echo \"ERROR: probe type is unknown: $PROBE_TYPE\" >/dev/stderr\n    exit $USAGE_ERR_CODE\n    ;;\n  esac\n\nelse\n  # anything else is a failing response. same behavior as Kubernetes' HTTP probe\n  echo \"FAIL: received an HTTP error code: $http_response\"\n  exit $PROBE_ERR_CODE\n\nfi\n" timed out
```
* Operator's logs, if necessary
* Crashing pod(s) logs, if necessary

  To get logs, use `kubectl -n <namespace> logs <pod name>`
  When pasting logs, always surround them with backticks or use the `insert code` button from the Github UI.
  Read [GitHub documentation if you need help](https://help.github.com/en/articles/creating-and-highlighting-code-blocks).

**Cluster Status to submit**:

* Output of kubectl commands, if necessary

  To get the health of the cluster, use `kubectl rook-ceph health`
  To get the status of the cluster, use `kubectl rook-ceph ceph status`
  For more details, see the [Rook kubectl Plugin](https://rook.io/docs/rook/latest-release/Troubleshooting/kubectl-plugin)

**Environment**:
* OS (e.g. from /etc/os-release): Ubuntu 22.04.4 LTS (Jammy)
* Kernel (e.g. `uname -a`): Linux svimtkump001 5.15.0-118-generic #128-Ubuntu SMP Fri Jul 5 09:28:59 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux
* Cloud provider or hardware configuration: running on Vmware
* Rook version (use `rook version` inside of a Rook Pod): v1.14.8
* Storage backend version (e.g. for ceph do `ceph -v`): ceph version 18.2.2 (531c0d11a1c5d39fbfe6aa8a521f023abf3bf3e2) reef (stable)
* Kubernetes version (use `kubectl version`): v1.30.3
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift): Kubernetes (plain vanilla)
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/rook/latest-release/Troubleshooting/ceph-toolbox/#interactive-toolbox)): HEALTH_OK

**Comment by BlaineEXE** [comment]
This sounds like the system is working as intended to me. As the RGW starts to consume more and more memory to service lots of clients, kubernetes may limit the memory it can use, which will mean the RGW can't service all of them. In that case, I would expect that the readiness probe would fail, indicating that the RGW is overloaded. Have you tried increasing the RGW memory limits above 1Gi?

**Comment by BlaineEXE** [comment]
> When memory limit is reached, or an oomkill should be done, or the pod process (radosgw) should crash, triggering a restart.

Kubernetes is in charge of OOM killing pods that overuse resource limits. To enable this, you have to enable resource `limits`. Resource `requests` will not enable K8s's OOM killing.

**Comment by TimVerboisCgk** [comment]
> This sounds like the system is working as intended to me. As the RGW starts to consume more and more memory to service lots of clients, kubernetes may limit the memory it can use, which will mean the RGW can't service all of them. In that case, I would expect that the readiness probe would fail, indicating that the RGW is overloaded. Have you tried increasing the RGW memory limits above 1Gi?

We can increase the memory limit and this will surely give us more capacity.  The problem I'm heaving with this is that it stops serving all together.  I would expect 2 possible outcomes:

1. The gateway reaches its memory limit and crashes (normal kubernetes behaviour), restarts and starts serving requests again
2. The gateway reaches its memory limit, goes into "not-ready", is lifted out of the kubernetes service and then the memory usage drops heavily since no new requests are coming in.  After memory drops below a certain point, the readiness probe starts working again and the gateway is returned into the service.

At this point, it does neither of these outcomes.  It hangs himself up because the readiness probe fails at a certain point, but the memory doesn't drop after that, it just stays in this hangup state forever.  

This means that when the service gets overloaded, 1 by 1, the pods go into not-ready rendering the service "down" completely until a human intervention restarts them all.  

This doesn't look normal behavior to me.

**Comment by TimVerboisCgk** [comment]
> Potentially this could describe a root cause for the issue. I wasn't able to find a clear explanation of this from upstream Ceph docs unfortunately: https://access.redhat.com/solutions/6727361

Well, this is a great tip.  Thanx!!

So, I checked the parameters:

```shell
[root@rook-ceph-tools-68976547-zz7tq /]# ceph config get client.rgw.applicationpool.a rgw_thread_pool_size
512
[root@rook-ceph-tools-68976547-zz7tq /]# ceph config get client.rgw.applicationpool.a rgw_max_concurrent_requests
1024
```
so, they are not the same.  We will change them to the same setting and we will test again.

Thanx again!

**Comment by TimVerboisCgk** [comment]
To conclude this issue, I can share what the problem was.  We had a problem with the number of object per shard.  There were way more object per shard then the advised max of 100.000.  We used the right commands to rearrange the objects over more shards in the bucket and once this was done, the problem with the gateways also disappeared.

So, it is clear to me now, that this problem was caused by a problem in the background and not by the gateways itself.
